### PR TITLE
[SPARK-41793][SQL] Incorrect result for window frames defined by a range clause on large decimals

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Use `DecimalAddNoOverflowCheck` instead of `Add` to craete bound ordering for window range frame

### Why are the changes needed?

Before 3.4, the `Add` did not check overflow. Instead, we always wrapped `Add` with a `CheckOverflow`. After https://github.com/apache/spark/pull/36698, we make `Add` check overflow by itself. However, the bound ordering of window range frame uses `Add` to calculate the boundary that is used to determine which input row lies within the frame boundaries of an output row. Then the behavior is changed with an extra overflow check. 

Technically，We could allow an overflowing value if it is just an intermediate result. So this pr use `DecimalAddNoOverflowCheck` to replace the `Add` to restore the previous behavior.

### Does this PR introduce _any_ user-facing change?

yes, restore the previous(before 3.4) behavior

### How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @tgravescs

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala [hunk 1]
```diff
@@ -474,4 +474,33 @@ class DataFrameWindowFramesSuite extends QueryTest with SharedSparkSession {
...
+    "decimals") {
+    withTempView("test_table") {
+      spark.sql(
+        """
+          |create or replace temp view test_table as
+          |select * from values
+          |  (1, cast('11342371013783243717493546650944543.47' as decimal(38,2))),
+          |  (1, cast('999999999999999999999999999999999999.99' as decimal(38,2)))
+          |as data(a, b);
+          |""".stripMargin)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala:488]: It seems simpler to use scala API for it
```
Seq(
  1 -> "113...",
  1 -> "999 ..."
).toDF("a", "b")
  .select("a", $"b".cast("decimal(38, 2)"))
  .createTempView("...")
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@ulysses-you to help other reviewers understand it, can you add more explaination in the PR description about how `boundExpr` is used in the window operator? To convince people that it's safe to skip overflow check for it.

**Review by gerashegalov — APPROVED** [review]
LGTM 

(verified that the failing spark-rapids test succeeds now)

**Comment by ulysses-you** [comment]
@cloud-fan addressed, hope it is helpful

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4!

**Pull request closed without merging**